### PR TITLE
[*-controller-manager]get rid of copied fields in the options and using componentconfig fields

### PR DESCRIPTION
**Pull request by stewart-yu**
**What type of PR is this?**
/kind cleanup
/kind feature

**What this PR does / why we need it**:

- There are so many coped fields in `kube-controller manager`( betwwen `config` and `option`), such as
 [`AttachDetachControllerOptions`](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/options/attachdetachcontroller.go#L27-L30) and [`AttachDetachControllerConfiguration`](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/apis/config/types.go#L166-L174), and so on.
- We should not maintain the copied fields. Get rid of copied fields in the options and directly point the flags to the componentconfig fields in this PR.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by sttts** [comment]
I would prefer to wrap the componentconfig types with an options struct using embedding. Then we can keep the current structure. Is that feasible?

**Comment by stewart-yu** [comment]
> I would prefer to wrap the componentconfig types with an options struct using embedding. Then we can keep the current structure. Is that feasible?

Yes, for now. but i have question:
Later, we should add `--config` for kube-controller manager, that is to say add two domain 
````
ComponentConfig kubectrlmrgconfig.KubeControllerManagerConfiguration
ConfigFile string
```` 
in `KubeControllerManagerOptions` struct at least, should we keep those option https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/options/options.go#L58-L80 in `KubeControllerManagerOptions` struct?

File: cmd/controller-manager/app/options/cloudprovider.go [hunk 1]
```diff
@@ -24,18 +24,18 @@ import (
...
+
 	return allErrors
 }
 
-// AddFlags adds flags related to cloudprovider for controller manager to the specified FlagSet.
-func (s *CloudProviderOptions) AddFlags(fs *pflag.FlagSet) {
+// AddCloudProviderConfigurationFlags adds flags related to cloudprovider for controller manager to the specified FlagSet.
+func AddCloudProviderConfigurationFlags(s *kubectrlmgrconfig.CloudProviderConfiguration, fs *pflag.FlagSet) {
 	fs.StringVar(&s.Name, "cloud-provider", s.Name,
 		"The provider for cloud services. Empty string for no provider.")
```
Comment (stewart-yu) [inline cmd/controller-manager/app/options/cloudprovider.go]: this function should find better place to hold, like the `LeaderElectionConfiguration` Flags. WDYT? @sttts

File: cmd/controller-manager/app/options/debugging.go [hunk 2]
```diff
@@ -24,12 +24,11 @@ import (
...
-	EnableProfiling           bool
-	EnableContentionProfiling bool
+	apiserverconfig.DebuggingConfiguration
 }
 
-// AddFlags adds flags related to debugging for controller manager to the specified FlagSet.
-func (o *DebuggingOptions) AddFlags(fs *pflag.FlagSet) {
+// AddDebuggingConfigurationFlags adds flags related to debugging for controller manager to the specified FlagSet.
+func AddDebuggingConfigurationFlags(o *apiserverconfig.DebuggingConfiguration, fs *pflag.FlagSet) {
 	if o == nil {
```
Comment (stewart-yu) [inline cmd/controller-manager/app/options/debugging.go]: The same

**Comment by sttts** [comment]
> in `KubeControllerManagerOptions` struct at least, should we keep those option https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/options/options.go#L58-L80 in `KubeControllerManagerOptions` struct?

This leads to the precedence question of flags vs. component config, doesn't it? IMO flags should override config values. In other words: we have to load the config and then apply flags to the structs of the config object. I think the kubelet has some (hacky) solution for that.

File: cmd/controller-manager/app/options/cloudprovider.go (hunk 1 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/cloudprovider.go]: may be as a follow-up

**Comment by stewart-yu** [comment]
cc @sttts  
have more comments about this PR?
I have confusion about how to fix it https://github.com/kubernetes/kubernetes/pull/70472#discussion_r229986479

**Comment by sttts** [comment]
I would like to see progress here. Sorry for not finding time for review earlier. Could you rebase the PR?

**Comment by stewart-yu** [comment]
/assign @sttts 
rebased, PTAL

File: cmd/controller-manager/app/options/generic.go [hunk 3]
```diff
@@ -30,29 +29,27 @@ import (
...
-	Port                    int32
-	Address                 string
-	MinResyncPeriod         metav1.Duration
-	ClientConnection        componentbaseconfig.ClientConnectionConfiguration
-	ControllerStartInterval metav1.Duration
-	LeaderElection          componentbaseconfig.LeaderElectionConfiguration
-	Debugging               *DebuggingOptions
-	Controllers             []string
+	*kubectrlmgrconfig.GenericControllerManagerConfiguration
+	DebuggingFlag *DebuggingOptions
```
Comment (sttts) [inline cmd/controller-manager/app/options/generic.go]: why not just `Debugging` ?

File: cmd/controller-manager/app/options/kubecloudshared.go [hunk 4]
```diff
@@ -19,39 +19,32 @@ package options
...
-	RouteReconciliationPeriod    metav1.Duration
-	NodeMonitorPeriod            metav1.Duration
-	ClusterName                  string
-	ClusterCIDR                  string
-	AllocateNodeCIDRs            bool
-	CIDRAllocatorType            string
-	ConfigureCloudRoutes         bool
-	NodeSyncPeriod               metav1.Duration
+	*kubectrlmgrconfig.KubeCloudSharedConfiguration
+	CloudProviderFlag *CloudProviderOptions
```
Comment (sttts) [inline cmd/controller-manager/app/options/kubecloudshared.go]: Why the `Flag` postfix?

File: cmd/controller-manager/app/options/generic.go (hunk 3 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/generic.go]: The origin thoughts is that do not override `Debugging` in `GenericControllerManagerConfiguration`. But seems override is OK.
fixed now

File: cmd/controller-manager/app/options/kubecloudshared.go (hunk 4 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/kubecloudshared.go]: fixed

File: cmd/controller-manager/app/options/generic.go [hunk 5]
```diff
@@ -30,29 +29,26 @@ import (
...
-	Debugging               *DebuggingOptions
-	Controllers             []string
+	*kubectrlmgrconfig.GenericControllerManagerConfiguration
+	Debugging *DebuggingOptions
 }
 
 // NewGenericControllerManagerConfigurationOptions returns generic configuration default values for both
 // the kube-controller-manager and the cloud-contoller-manager. Any common changes should
 // be made here. Any individual changes should be made in that controller.
 func NewGenericControllerManagerConfigurationOptions(cfg kubectrlmgrconfig.GenericControllerManagerConfiguration) *GenericControllerManagerConfigurationOptions {
```
[the remark this answers is not in the document]
Comment (stewart-yu) [reply cmd/controller-manager/app/options/generic.go]: fixed

**Review by luxas — COMMENTED** [review]
Thanks @stewart-yu, very needed cleanup!!

**Pull request merged**